File: solutions/python/robot-simulator/robot_simulator.py
"""A robot factory's test facility program to verify robot movements."""

# Globals for the directions used as indices when accessing movement lists
NORTH = 0
EAST = 1
SOUTH = 2
WEST = 3

TURN_RIGHT = [EAST, SOUTH, WEST, NORTH]
TURN_LEFT = [WEST, NORTH, EAST, SOUTH]
ADVANCE = [(0, 1), (1, 0), (0, -1), (-1, 0)]


class Robot:
    """Robots are placed on a hypothetical infinite grid, facing a particular
    direction (north, east, south, or west) at a set of {x,y} coordinates"""

    def __init__(self, direction=NORTH, x_pos=0, y_pos=0):
        self.direction = direction
        self.coordinates = (x_pos, y_pos)

    def move(self, instructions: str):
        """The robots have three possible movements:
        R = turn right, L = turn left, A = advance"""
        for inst in instructions:
            match inst:
                case "R":
                    self.direction = TURN_RIGHT[self.direction]
                case "L":
                    self.direction = TURN_LEFT[self.direction]
                case "A":
                    self.coordinates = tuple(
                        map(sum, zip(self.coordinates, ADVANCE[self.direction]))
                    )

File: solutions/javascript/robot-simulator/robot-simulator.js
//
// A robot factory's test facility program to verify robot movements.
//
export class InvalidInputError extends Error {
  constructor(message) {
    super();
    this.message = message || "Invalid Input";
  }
}

const DIRECTIONS = ["north", "east", "south", "west"];

export class Robot {
  #facing = "north";
  #x = 0;
  #y = 0;

  get bearing() {
    return this.#facing;
  }

  get coordinates() {
    return [this.#x, this.#y];
  }

  place({ x, y, direction }) {
    if (!DIRECTIONS.includes(direction)) {
      throw new InvalidInputError("Invalid direction!");
    }
    this.#facing = direction;
    this.#x = x;
    this.#y = y;
  }

  evaluate(instructions) {
    [...instructions].forEach((inst) => {
      switch (inst) {
        case "R":
          this.#turn_right();
          break;
        case "L":
          this.#turn_left();
          break;
        case "A":
          this.#advance();
          break;
      }
    });
  }

  #turn_right() {
    switch (this.#facing) {
      case "north":
        this.#facing = "east";
        break;
      case "east":
        this.#facing = "south";
        break;
      case "south":
        this.#facing = "west";
        break;
      case "west":
        this.#facing = "north";
        break;
    }
  }

  #turn_left() {
    switch (this.#facing) {
      case "north":
        this.#facing = "west";
        break;
      case "east":
        this.#facing = "north";
        break;
      case "south":
        this.#facing = "east";
        break;
      case "west":
        this.#facing = "south";
        break;
    }
  }

  #advance() {
    switch (this.#facing) {
      case "north":
        this.#y += 1;
        break;
      case "east":
        this.#x += 1;
        break;
      case "south":
        this.#y -= 1;
        break;
      case "west":
        this.#x -= 1;
        break;
    }
  }
}
